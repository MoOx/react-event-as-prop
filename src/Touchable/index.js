// @flow

import React, { Component } from "react"

type PropsType = {
  onMouseDown?: Function | boolean,
  onMouseUp?: Function | boolean,
}

type StateType = {
  touched: boolean,
}

// @todo try a more complex type for ComposedComponent + return type
export default (ComposedComponent: Function) => {
  class Touchable extends Component<void, PropsType, StateType> {
    props: PropsType

    state: StateType = {
      touched: false,
    }

    handleMouseDown: Function = (event: SyntheticEvent): void => {
      this.setState({ touched: true })
      if (typeof this.props.onMouseDown === "function") {
        this.props.onMouseDown(event)
      }
    }

    handleMouseUp: Function = (event: SyntheticEvent): void => {
      this.setState({ touched: false })
      if (typeof this.props.onMouseUp === "function") {
        this.props.onMouseUp(event)
      }
    }

    render(): React$Element<any> {
      return (
        <ComposedComponent
          { ...this.props }
          { ...this.state }
          onMouseDown={ this.handleMouseDown }
          onMouseUp={ this.handleMouseUp }
        />
      )
    }
  }

  return Touchable
}
