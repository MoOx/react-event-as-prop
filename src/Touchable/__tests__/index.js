import React from "react"
import renderer from "react-test-renderer"

import Touchable from ".."

const Compo = ({ touched, ...props }: { touched: boolean }) => (
  <span { ...props }>{ touched ? "touched" : "nope" }</span>
)
const Composed = Touchable(Compo)

test("touched props changes when touched", () => {
  const component = renderer.create(
    <Composed
      test
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  tree.props.onMouseDown()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  tree.props.onMouseUp()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

let flag = 0
const handleMouseDown = () => flag = 1
const handleMouseUp = () => flag = 2

test("touched props changes when touched and execute on{Event}", () => {
  const component = renderer.create(
    <Composed
      test
      onMouseDown={ handleMouseDown }
      onMouseUp={ handleMouseUp }
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(flag).toEqual(0)

  tree.props.onMouseDown()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(flag).toEqual(1)

  tree.props.onMouseUp()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(flag).toEqual(2)
})
