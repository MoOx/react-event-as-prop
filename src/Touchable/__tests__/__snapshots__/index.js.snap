exports[`test touched props changes when touched 1`] = `
<span
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  test={true}>
  nope
</span>
`;

exports[`test touched props changes when touched 2`] = `
<span
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  test={true}>
  touched
</span>
`;

exports[`test touched props changes when touched 3`] = `
<span
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  test={true}>
  nope
</span>
`;

exports[`test touched props changes when touched and execute on{Event} 1`] = `
<span
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  test={true}>
  nope
</span>
`;

exports[`test touched props changes when touched and execute on{Event} 2`] = `
<span
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  test={true}>
  touched
</span>
`;

exports[`test touched props changes when touched and execute on{Event} 3`] = `
<span
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  test={true}>
  nope
</span>
`;
